pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *PrivilegedAccessManagerEntitlementIdentity
+	gcpClient *gcp.Client
+	desired   *krm.PrivilegedAccessManagerEntitlement
+	actual    *privilegedaccessmanagerpb.Entitlement
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+
+	req := &privilegedaccessmanagerpb.GetEntitlementRequest{Name: a.id.FullyQualifiedName()}
+	entitlementpb, err := a.gcpClient.GetEntitlement(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("error getting PrivilegedAccessManagerEntitlement %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = entitlementpb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := PrivilegedAccessManagerEntitlementSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &privilegedaccessmanagerpb.CreateEntitlementRequest{
+		Parent:        a.id.Parent.String(),
+		EntitlementId: a.id.Entitlement,
+		Entitlement:   resource,
+	}
+	op, err := a.gcpClient.CreateEntitlement(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error creating PrivilegedAccessManagerEntitlement %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("error waiting for PrivilegedAccessManagerEntitlement %s to be created: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+
+	status := &krm.PrivilegedAccessManagerEntitlementStatus{}
+	observedState := PrivilegedAccessManagerEntitlementStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ObservedState = observedState
+	status.ExternalRef = a.id.AsExternalRef()
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(AdditionalNotificationTargets_FromProto(mapCtx, a.actual.AdditionalNotificationTargets), a.desired.Spec.AdditionalNotificationTargets) {
```

Comment:
Nit: It should be more concise to call  PrivilegedAccessManagerEntitlementSpec_FromProto and then check the fields, rather than mapping each field individually.  I think you can also call PrivilegedAccessManagerEntitlementSpec_ToProto and check the proto fields.

Reply:
Will refactor it in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2659.